server/attr/attributeContext.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package attr
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+
+	mixerpb "istio.io/mixer/api/v1"
+)
+
+// This code would be quite a bit nicer with generics... Oh well.
+
+// AttributeContext maintains an independent set of attributes.
+type AttributeContext interface {
+	// GetString returns the named attribute if it exists.
+	GetString(name string) (string, bool)
+
+	// GetInt64 returns the named attribute if it exists.
+	GetInt64(name string) (int64, bool)
+
+	// GetFloat64 returns the named attribute if it exists.
+	GetFloat64(name string) (float64, bool)
+
+	// GetBool returns the named attribute if it exists.
+	GetBool(name string) (bool, bool)
+
+	// GetTime returns the named attribute if it exists.
+	GetTime(name string) (time.Time, bool)
+
+	// GetBytes returns the named attribute if it exists.
+	GetBytes(name string) ([]uint8, bool)
+}
+
+type attributeContext struct {
+	stringAttributes  map[string]string
+	int64Attributes   map[string]int64
+	float64Attributes map[string]float64
+	boolAttributes    map[string]bool
+	timeAttributes    map[string]time.Time
+	bytesAttributes   map[string][]uint8
+}
+
+func newAttributeContext() *attributeContext {
+	ac := &attributeContext{}
+	ac.reset()
+	return ac
+}
+
+// Ensure that all dictionary indices are valid and that all values
+// are in range.
+//
+// Note that since we don't have the attribute schema, this doesn't validate
+// that a given attribute is being treated as the right type. That is, an
+// attribute called 'source.ip' which is of type IP_ADDRESS could be listed as
+// a string or an int, and we wouldn't catch it here.
+func checkPreconditions(dictionary dictionary, attrs *mixerpb.Attributes) error {
+	for k := range attrs.StringAttributes {
+		_, present := dictionary[k]
+		if !present {
+			return fmt.Errorf("attribute index %d is not defined in the current dictionary", k)
```

Comment:
do we expect anything that uses checkPreconditions to be able to handle this error specifically?  Should we consider exporting a var for AttributeNotFound errors?